File: src/main/java/fit/nlu/tmdt/modules/post/service/PostService.java
package fit.nlu.tmdt.modules.post.service;

import fit.nlu.tmdt.modules.post.dto.request.CreatePostRequest;
import fit.nlu.tmdt.modules.post.dto.request.PostSearchParams;
import fit.nlu.tmdt.modules.post.dto.request.UpdatePostRequest;
import fit.nlu.tmdt.modules.post.dto.response.PostResponse;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;
import java.util.Map;

/**
 * Post Service Interface
 */
public interface PostService {

    /**
     * Search posts với filter
     */
    Page<PostResponse> searchPosts(PostSearchParams params, Pageable pageable, Long userId);

    /**
     * Get post by ID
     */
    PostResponse getPostById(Long id, Long userId);

    /**
     * Create new post
     */
    PostResponse createPost(CreatePostRequest request, Long landlordId);

    /**
     * Update post
     */
    PostResponse updatePost(Long id, UpdatePostRequest request, Long landlordId);

    /**
     * Delete post (soft delete)
     */
    void deletePost(Long id, Long landlordId);

    /**
     * Get my posts
     */
    Page<PostResponse> getMyPosts(Long landlordId, Pageable pageable);

    /**
     * Get post stats
     */
    PostStatsResponse getPostStats(Long postId, Long landlordId);

    /**
     * Check if user is owner
     */
    boolean isOwner(Long postId, Long userId);

    /**
     * Increment view count async
     */
    void incrementViewCountAsync(Long postId);

    /**
     * Admin: Approve post
     */
    void approvePost(Long postId, Long adminId);

    /**
     * Admin: Reject post
     */
    void rejectPost(Long postId, String reason, Long adminId);

    /**
     * Admin: Get pending posts
     */
    Page<PostResponse> getPendingPosts(Pageable pageable);

    /**
     * Boost a post
     */
    Map<String, Object> boostPost(Long postId, Long boostPackageId, Long landlordId);

    /**
     * Extend post expiration
     */
    Map<String, Object> extendPost(Long postId, int days, Long landlordId);

    /**
     * Search public posts (no auth required)
     */
    Page<PostResponse> searchPublicPosts(PostSearchParams params, Pageable pageable);

    /**
     * Get featured posts (no auth required)
     */
    List<PostResponse> getFeaturedPosts(int limit);

    record PostStatsResponse(
            Long postId,
            Integer viewCount,
            Integer favoriteCount,
            Integer contactCount,
            Integer bookingCount,
            Integer completedBookings
    ) {}
}

File: frontend/src/services/postService.ts
import api from './api';
import type {
  ApiResponse,
  PaginatedData,
  PostResponse,
  PostStatsResponse,
  LandlordDashboardStats,
} from '@/types';

export interface PostSearchParams {
  keyword?: string;
  minPrice?: number;
  maxPrice?: number;
  areaMin?: number;
  areaMax?: number;
  location?: string;
  district?: string;
  province?: string;
  ward?: string;
  amenityIds?: number[];
  category?: string;
  latitude?: number;
  longitude?: number;
  radiusKm?: number;
  nearbyUniversityId?: number;
  petFriendly?: boolean;
  parkingAvailable?: boolean;
  hasBalcony?: boolean;
  hasWindows?: boolean;
}

export interface AdminPostQuery extends PostSearchParams {
  status?: string;
  page?: number;
  size?: number;
  sortBy?: string;
  sortDirection?: string;
}

export const postService = {
  /**
   * Get all public posts (Search & List)
   */
  getPublicPosts: async (params: PostSearchParams & { page?: number; size?: number; sortBy?: string; sortDirection?: string }): Promise<PaginatedData<PostResponse>> => {
    const response = await api.get<ApiResponse<PaginatedData<PostResponse>>>('/v1/posts/public', {
      params,
    });
    return response.data.data!;
  },

  /**
   * Get featured posts for home page
   */
  getFeaturedPosts: async (limit: number = 10): Promise<PostResponse[]> => {
    const response = await api.get<ApiResponse<PostResponse[]>>('/v1/posts/featured', {
      params: { limit },
    });
    return response.data.data!;
  },

  /**
   * Get post details by ID
   */
  getPostById: async (id: number | string): Promise<PostResponse> => {
    const response = await api.get<ApiResponse<PostResponse>>(`/v1/posts/${id}`);
    return response.data.data!;
  },

  /**
   * Get landlord's own posts
   */
  getMyPosts: async (page: number = 0, size: number = 10): Promise<PaginatedData<PostResponse>> => {
    const response = await api.get<ApiResponse<PaginatedData<PostResponse>>>('/v1/posts/my', {
      params: { page, size },
    });
    return response.data.data!;
  },

  /**
   * Admin: Get posts for moderation
   */
  getAdminPosts: async (params: AdminPostQuery = {}): Promise<PaginatedData<PostResponse>> => {
    const response = await api.get<ApiResponse<PaginatedData<PostResponse>>>('/v1/posts/admin', {
      params,
    });
    return response.data.data!;
  },

  /**
   * Admin: Approve a post
   */
  approvePost: async (id: number | string): Promise<void> => {
    await api.put(`/v1/posts/${id}/approve`);
  },

  /**
   * Admin: Reject a post
   */
  rejectPost: async (id: number | string, reason: string): Promise<void> => {
    await api.put(`/v1/posts/${id}/reject`, { reason });
  },

  /**
   * Create a new post (Landlord only)
   */
  createPost: async (data: any): Promise<PostResponse> => {
    const response = await api.post<ApiResponse<PostResponse>>('/v1/posts', data);
    return response.data.data!;
  },

  /**
   * Update an existing post (Landlord only)
   */
  updatePost: async (id: number | string, data: any): Promise<PostResponse> => {
    const response = await api.put<ApiResponse<PostResponse>>(`/v1/posts/${id}`, data);
    return response.data.data!;
  },

  /**
   * Delete a post (Landlord only)
   */
  deletePost: async (id: number | string): Promise<void> => {
    await api.delete(`/v1/posts/${id}`);
  },

  /**
   * Boost a post
   */
  boostPost: async (id: number | string, boostPackageId: number): Promise<any> => {
    const response = await api.post<ApiResponse<any>>(`/v1/posts/${id}/boost`, null, {
      params: { boostPackageId },
    });
    return response.data.data;
  },

  /**
   * Extend post expiration
   */
  extendPost: async (id: number | string, days: number = 30): Promise<any> => {
    const response = await api.post<ApiResponse<any>>(`/v1/posts/${id}/extend`, null, {
      params: { days },
    });
    return response.data.data;
  },

  /**
   * Get post statistics
   */
  getPostStats: async (id: number | string): Promise<PostStatsResponse> => {
    const response = await api.get<ApiResponse<PostStatsResponse>>(`/v1/posts/${id}/stats`);
    return response.data.data!;
  },

  /**
   * Get landlord dashboard statistics
   */
  getDashboardStats: async (): Promise<LandlordDashboardStats> => {
    const response = await api.get<ApiResponse<LandlordDashboardStats>>('/v1/posts/landlord/dashboard/stats');
    return response.data.data!;
  },

  /**
   * Record post contact (when user clicks contact buttons)
   */
  recordContact: async (id: number | string): Promise<void> => {
    await api.post(`/v1/posts/${id}/contact`);
  },
};

export default postService;
